File: api/src/main/java/com/aa/weatherprediction/service/ForecastService.java

package com.aa.weatherprediction.service;

import com.aa.weatherprediction.model.DayAndReport;
import com.aa.weatherprediction.model.Report;
import com.aa.weatherprediction.model.WeatherData;
import com.aa.weatherprediction.model.City;
import com.aa.weatherprediction.model.Weather;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ForecastService {

    private final ConnectorService connectorService;

    public ForecastService(ConnectorService connectorService) {
        this.connectorService = connectorService;
    }

    public Report getForecastByCity(String name) {
        City city = connectorService.getCity(name).get(0);
        WeatherData weatherData = connectorService.getWeather(city.getLat(), city.getLon());
        List<String> weatherAlerts = new ArrayList<>();
        if (weatherData.getWind().getSpeed() > 10) {
            weatherAlerts.add("It’s too windy, watch out!");
        }
        if (weatherData.getMain().getTemp() > 40) {
            weatherAlerts.add("Use sunscreen lotion");
        }
        if (weatherData.getRain() != null) {
            weatherAlerts.add("Carry Umbrella");
        }
        return new Report(
                weatherAlerts,
                weatherData.getWeather().stream().map(Weather::getMain).collect(Collectors.toList()),
                weatherData.getMain().getTempMin(),
                weatherData.getMain().getTempMax()
        );
    }

    public List<DayAndReport> getForecastByDays(String name, long days) {
        List<DayAndReport> reports = new ArrayList<>();
        for (long day = 1; day <= days; day++) {
            reports.add(new DayAndReport(
                    LocalDate.now().plusDays(day).getDayOfWeek().name(),
                    getForecastByCity(name)
            ));
        }
        return reports;
    }
}

File: weather-prediction-api/src/main/kotlin/com/aa/weatherprediction/service/ForecastService.kt
package com.aa.weatherprediction.service

import com.aa.weatherprediction.model.*
import org.springframework.messaging.Message
import org.springframework.beans.factory.annotation.Value
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.support.KafkaHeaders
import org.springframework.messaging.support.MessageBuilder
import org.springframework.stereotype.Service
import java.time.LocalDate
import kotlin.collections.ArrayList

@Service
class ForecastService(
    private val kafkaTemplate: KafkaTemplate<String, Any>,
    private val connectorService: ConnectorService
) {

    @Value("\${kafka.topic}")
    private val topic: String = ""

    fun getForecastByCity(name: String): Report {
        sendEvent(EVENT.ORDER)
        val city = connectorService.getCity(name).first()
        val weatherData: WeatherData = connectorService.getWeather(city.lat!!, city.lon!!)
        val weatherAlerts: ArrayList<String> = arrayListOf()
        if (weatherData.wind?.speed!! > 10) {
            weatherAlerts.add("It’s too windy, watch out!")
        }
        if (weatherData.main?.temp!! > 40) {
            weatherAlerts.add("Use sunscreen lotion")
        }
        if (weatherData.rain != null) {
            weatherAlerts.add("Carry Umbrella")
        }
        sendEvent(EVENT.SUCCESS)
        return Report(
            alerts = weatherAlerts,
            mains = weatherData.weather?.map { it.main!! },
            temp_min = weatherData.main?.temp_min,
            temp_max = weatherData.main?.temp_max,
        )
    }

    private fun sendEvent(type: EVENT) {
        val eventMessage: Message<UsagePrice> = MessageBuilder
            .withPayload(UsagePrice("User1", type))
            .setHeader(KafkaHeaders.TOPIC, topic)
            .setHeader("X-Custom-Header", "Custom header here")
            .build()
        kafkaTemplate.send(eventMessage)
    }

    // using Mock Data as forecast service comes under Pro subscription
    fun getForecastByDays(name: String, days: Long): ArrayList<DayAndReport> {
        var reports = arrayListOf<DayAndReport>()
        for(day in 1..days) {
            reports.add(
                DayAndReport(
                    day = LocalDate.now().plusDays(day).dayOfWeek.name,
                    report = getForecastByCity(name)
                )
            )
        }
        return reports
    }
}
